Allow meters to be made invisible to unauthenticated users

NoraCodes opened the pull request:
This is a WIP pull request. I am interested in feedback but it is not ready to be merged.

This PR adds the ability to turn off the display of meters, and prevents unauthenticated users from seeing non-displayable meters. It also adds a meter management UI.

NoraCodes committed:
Add dual toggle for meters.

Meters can now be enabled or not (collecting data) and visible or not
(shown to users). There are also methods (and tests) to get only the
meters marked with the appropriate flags.

Meters are enabled and visible by default.

NoraCodes committed:
Use only enabled meters in updateMamacMeters.

NoraCodes committed:
Filter meters by authentication status

Now, non-authenticated users can't see non-displayable meters

NoraCodes committed:
UI for meters toggling; display only

Provides a tabular view of all meters in the system, with all associated
information (including toggles). Users who are not logged in see only
meters with the Displayable flag on, regardless of what view is in use,
and the server does not send IP or meter type in the absence of a token.

NoraCodes committed:
Fix TSC/ATLoader not giving errors in some cases

ATLoader needs to have a plugin enabled to give good error messages and
we hadn't done that, so here it is.

shuheif commented on src/client/app/components/HeaderButtonsComponent.tsx:
```diff
@@ -24,33 +24,14 @@ export default class HeaderButtonsComponent extends React.Component<HeaderButton
+		const renderLoginButton = !hasToken();
+		const renderHomeButton = getPage() !== '';
+		const renderAdminButton = hasToken() && getPage() !== 'admin';
+		const renderGroupsButton = hasToken() && getPage() !== 'groups';
```
I think we don't need the hasToken() check here as the current development branch allows rendering the button without logging in.

NoraCodes replied:
Like this?

```suggestion
		const renderGroupsButton = getPage() !== 'groups';
```

shuheif replied:
Yes

shuheif commented:
This new feature is cool! I think the structure of both the front-end and the back-end makes sense and looks good to me. As long as I see, there are 
- 'column "displayable" does not exist' error when adding new meters, and 
- the implementation for actual enabled/disabled switch button

 to be done. Do you have any concerns or anything that you want to discuss about at this point @LeoTindall?

NoraCodes commented:
> 'column "displayable" does not exist' error when adding new meters, and

I don't get this error; did you run the database migration?

> the implementation for actual enabled/disabled switch button

This is on hold until Obvius stuff is done; once I get time to work on this, I would definitely appreciate your input.

NoraCodes committed:
Add buttons for editing meter toggles. Not active.

NoraCodes committed:
Meter editing from the interface.

Adds new actions and improves error handling for authentication.

NoraCodes commented:
This pull request is no longer WIP and requires review

shuheif commented:
The code looks good in general. Although the Travis test didn't work, the code passes all the tests on my laptop. I noticed a few things:

- I think the Group button on the header should be visible to the normal users. (As mentioned before)

- The text of "Save meter edits" button is cut off in the middle.

- I think it would be nicer to hide the save button to non-admin users if possible.

- What happens if a group contains disabled/hidden meters? I couldn't check this because the add-group feature is currently broken, also I couldn't find the code to support groups. I'm sorry if this is out of scope for this PR.

NoraCodes commented:
> The code looks good in general. Although the Travis test didn't work, the code passes all the tests on my laptop. I noticed a few things:
> 
>     * I think the Group button on the header should be visible to the normal users. (As mentioned before)

Definitely, I forgot that I overwrote that. I'll fix it.

>     * The text of "Save meter edits" button is cut off in the middle.

Can you screenshot this? I can't repro.

>     * I think it would be nicer to hide the save button to non-admin users if possible.

Definitely, didn't even think of that but you're right.

>     * What happens if a group contains disabled/hidden meters? I couldn't check this because the add-group feature is currently broken, also I couldn't find the code to support groups. I'm sorry if this is out of scope for this PR.

The data is still included in the group. This is useful to admins so they can group meters together which are nonsensical by themselves.

shuheif commented:
Here're the screenshots. Apparently, the save button shrinks as the screen gets smaller. I checked it on Google Chrome and Safari.
<img width="1280" alt="screen shot 2018-11-15 at 12 23 33 am" src="https://user-images.githubusercontent.com/10185165/48534233-12d79e00-e86d-11e8-9576-de1834264c02.png">
<img width="1280" alt="screen shot 2018-11-15 at 12 23 58 am" src="https://user-images.githubusercontent.com/10185165/48534234-12d79e00-e86d-11e8-87e2-ed5e2bfae4de.png">

NoraCodes committed:
Fix for groups button missing

NoraCodes committed:
Merge development to meter-update-display-optional

NoraCodes committed:
Fix meter edit button getting squished

NoraCodes committed:
Improve meters list for unauthenticated users.

Fix button size, remove button and adming columns for unauthed users.

NoraCodes committed:
Fix linting failures.

NoraCodes committed:
Update meters-update-display-optional<-development

Including reconciling package-lock.json, 0.3.0 to 0.4.0 migration, and
making the baseline tests use the right Meter constructor.

NoraCodes committed:
Merge branch 'development' of github.com:OpenEnergyDashboard/OED into meter-update-display-optional
